### ISSUE TITLE
1.2.20 segfault

### ISSUE THREAD

**Issue by banks** [acassen/keepalived#359]
I have keepalived setup using unicast between two AWS nodes in same subnet.

One becomes master and is fine, the other becomes backup but then gets into a segfault loop and continually crashes and restarts with:

```
Jun 16 10:58:09 prometheus Keepalived[26642]: Starting Keepalived v1.2.20 (05/04,2016)
Jun 16 10:58:09 prometheus Keepalived[26642]: Starting Healthcheck child process, pid=26644
Jun 16 10:58:09 prometheus Keepalived[26642]: Starting VRRP child process, pid=26645
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Registering Kernel netlink reflector
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Registering Kernel netlink command channel
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Opening file '/etc/keepalived/keepalived.conf'.
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Opening file '/etc/keepalived/conf.d/00_keepalived_vrrp_script__prometheus_http_check__.conf'.
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Opening file '/etc/keepalived/conf.d/keepalived_vrrp_instance__prometheus__.conf'.
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Registering Kernel netlink reflector
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Registering Kernel netlink command channel
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Registering gratuitous ARP shared channel
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Opening file '/etc/keepalived/keepalived.conf'.
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Opening file '/etc/keepalived/conf.d/00_keepalived_vrrp_script__prometheus_http_check__.conf'.
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Opening file '/etc/keepalived/conf.d/keepalived_vrrp_instance__prometheus__.conf'.
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Truncating auth_pass to 8 characters
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: Using LinkWatch kernel netlink reflector...
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: VRRP_Instance(prometheus) Entering BACKUP STATE
Jun 16 10:58:09 prometheus Keepalived_healthcheckers[26644]: Using LinkWatch kernel netlink reflector...
Jun 16 10:58:09 prometheus Keepalived_vrrp[26645]: VRRP_Script(prometheus_http_check) succeeded
Jun 16 10:58:11 prometheus kernel: keepalived[26645]: segfault at 10 ip 00000000004193a8 sp 00007ffe5623f690 error 4 in keepalived[400000+37000]
Jun 16 10:58:10 prometheus Keepalived[26642]: pid 26645 exited due to segmentation fault (SIGSEGV).
Jun 16 10:58:10 prometheus Keepalived[26642]:   Please report a bug at https://github.com/acassen/keepalived/issues
Jun 16 10:58:10 prometheus Keepalived[26642]:   and include this log from when keepalived started, what happened
Jun 16 10:58:10 prometheus Keepalived[26642]:   immediately before the crash, and your configuration file.
Jun 16 10:58:10 prometheus Keepalived[26642]: VRRP child process(26645) died: Respawning
Jun 16 10:58:10 prometheus Keepalived[26642]: Starting VRRP child process, pid=26661
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Registering Kernel netlink reflector
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Registering Kernel netlink command channel
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Registering gratuitous ARP shared channel
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Opening file '/etc/keepalived/keepalived.conf'.
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Opening file '/etc/keepalived/conf.d/00_keepalived_vrrp_script__prometheus_http_check__.conf'.
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Opening file '/etc/keepalived/conf.d/keepalived_vrrp_instance__prometheus__.conf'.
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Truncating auth_pass to 8 characters
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: Using LinkWatch kernel netlink reflector...
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: VRRP_Instance(prometheus) Entering BACKUP STATE
Jun 16 10:58:10 prometheus Keepalived_vrrp[26661]: VRRP_Script(prometheus_http_check) succeeded
Jun 16 10:58:12 prometheus kernel: keepalived[26661]: segfault at 10 ip 00000000004193a8 sp 00007ffe5623f620 error 4 in keepalived[400000+37000]
Jun 16 10:58:11 prometheus Keepalived[26642]: pid 26661 exited due to segmentation fault (SIGSEGV).
Jun 16 10:58:11 prometheus Keepalived[26642]:   Please report a bug at https://github.com/acassen/keepalived/issues
Jun 16 10:58:11 prometheus Keepalived[26642]:   and include this log from when keepalived started, what happened
Jun 16 10:58:11 prometheus Keepalived[26642]:   immediately before the crash, and your configuration file.
Jun 16 10:58:11 prometheus Keepalived[26642]: VRRP child process(26661) died: Respawning
[ repeats ]
```

Config is (apologies for the formatting, blame Chef):

```
vrrp_script prometheus_http_check {
	script /opt/prometheus/check_prometheus.sh
	interval 5
	fall 2
	rise 2
	}
vrrp_instance prometheus {
	notify_master /opt/prometheus/failover.sh
	virtual_router_id 42
	state BACKUP
	interface eth0
	unicast_src_ip 172.16.50.113
	unicast_peer {
		172.16.50.111
		}
	priority 99
	authentication {
		auth_type PASS
		auth_pass REDACTED
		}
	track_script {
		prometheus_http_check
		}
	nopreempt
	}
```

The other host is identical but with unicast src and peer up reversed and with priority 100

**Comment by banks** [comment]
Each time I restart keepalived on one host, the other host that was stuck in segfault loop becomes master (naturally) and then works just fine and the one restart comes back up in segfault loop like above.

**Comment by banks** [comment]
Downgrading to 1.2.13 fixed this. FWIW.

**Comment by pqarmitage** [comment]
Many thanks for the report.

It looks as though this could be the issue which was resolved in commit 36beb9e, where keepalived segfaults if no VIPS are configured. That commit was new in 1.2.21, so upgrading to 1.2.21 or later (1.2.22 is the current version) may resolve your problem.

If it doesn't resolve your problem, then could you try the following with 1.2.20 (or better still 1.2.22):

1. Build keepalived with debugging information: make debug
2. Run keepalived with the -m option. This should produce core dumps in /, and you will probably want to delete most of them.
3. Run gdb as: gdb <path to debug keepalived> /core.nnnn
4. Generate a stack trace by entering bt at the gdb prompt
5. Post the details of the stack trace in this issue report.

I  will then have a look at the stack trace and we should be able to work out what is happening.
